File: agenkit/adapters/python/codec.py
"""Message serialization and deserialization for protocol adapter."""

import json
from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from agenkit.interfaces import Message, ToolResult

from .errors import InvalidMessageError, MalformedPayloadError, UnsupportedVersionError

PROTOCOL_VERSION = "1.0"


def encode_message(message: Message) -> dict[str, Any]:
    """Encode a Message object to a dictionary for JSON serialization.

    Args:
        message: The Message to encode

    Returns:
        Dictionary representation of the message
    """
    return {
        "role": message.role,
        "content": message.content,
        "metadata": message.metadata,
        "timestamp": message.timestamp.isoformat(),
    }


def decode_message(data: dict[str, Any]) -> Message:
    """Decode a dictionary to a Message object.

    Args:
        data: Dictionary representation of a message

    Returns:
        Decoded Message object

    Raises:
        MalformedPayloadError: If message data is invalid
    """
    try:
        return Message(
            role=data["role"],
            content=data["content"],
            metadata=data.get("metadata", {}),
            # Parse timestamp if present, otherwise use current time
            timestamp=(
                datetime.fromisoformat(data["timestamp"])
                if "timestamp" in data
                else datetime.now(timezone.utc)
            ),
        )
    except (KeyError, ValueError, TypeError) as e:
        raise MalformedPayloadError(f"Failed to decode message: {e}", {"data": data}) from e


def encode_tool_result(result: ToolResult) -> dict[str, Any]:
    """Encode a ToolResult object to a dictionary for JSON serialization.

    Args:
        result: The ToolResult to encode

    Returns:
        Dictionary representation of the tool result
    """
    return {
        "success": result.success,
        "data": result.data,
        "error": result.error,
        "metadata": result.metadata,
    }


def decode_tool_result(data: dict[str, Any]) -> ToolResult:
    """Decode a dictionary to a ToolResult object.

    Args:
        data: Dictionary representation of a tool result

    Returns:
        Decoded ToolResult object

    Raises:
        MalformedPayloadError: If tool result data is invalid
    """
    try:
        return ToolResult(
            success=data["success"],
            data=data.get("data"),
            error=data.get("error"),
            metadata=data.get("metadata", {}),
        )
    except (KeyError, TypeError) as e:
        raise MalformedPayloadError(f"Failed to decode tool result: {e}", {"data": data}) from e


def create_request_envelope(
    method: str, agent_name: str | None = None, payload: dict[str, Any] | None = None
) -> dict[str, Any]:
    """Create a protocol request envelope.

    Args:
        method: Method name ("process" or "execute")
        agent_name: Name of the target agent (for agent requests)
        payload: Request payload

    Returns:
        Request envelope dictionary
    """
    return {
        "version": PROTOCOL_VERSION,
        "type": "request",
        "id": str(uuid4()),
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "payload": {
            "method": method,
            **({"agent_name": agent_name} if agent_name else {}),
            **(payload or {}),
        },
    }


def create_response_envelope(request_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    """Create a protocol response envelope.

    Args:
        request_id: ID of the request being responded to
        payload: Response payload

    Returns:
        Response envelope dictionary
    """
    return {
        "version": PROTOCOL_VERSION,
        "type": "response",
        "id": request_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "payload": payload,
    }


def create_error_envelope(
    request_id: str,
    error_code: str,
    error_message: str,
    error_details: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Create a protocol error envelope.

    Args:
        request_id: ID of the request that failed
        error_code: Error code
        error_message: Human-readable error message
        error_details: Additional error details

    Returns:
        Error envelope dictionary
    """
    return {
        "version": PROTOCOL_VERSION,
        "type": "error",
        "id": request_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "payload": {
            "error_code": error_code,
            "error_message": error_message,
            "error_details": error_details or {},
        },
    }


def create_stream_chunk_envelope(request_id: str, message: dict[str, Any]) -> dict[str, Any]:
    """Create a protocol stream chunk envelope.

    Args:
        request_id: ID of the streaming request
        message: Message chunk payload

    Returns:
        Stream chunk envelope dictionary
    """
    return {
        "version": PROTOCOL_VERSION,
        "type": "stream_chunk",
        "id": request_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "payload": {"message": message},
    }


def create_stream_end_envelope(request_id: str) -> dict[str, Any]:
    """Create a protocol stream end envelope.

    Args:
        request_id: ID of the streaming request

    Returns:
        Stream end envelope dictionary
    """
    return {
        "version": PROTOCOL_VERSION,
        "type": "stream_end",
        "id": request_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "payload": {},
    }


def validate_envelope(envelope: dict[str, Any]) -> None:
    """Validate a protocol envelope.

    Args:
        envelope: Envelope to validate

    Raises:
        InvalidMessageError: If envelope is invalid
        UnsupportedVersionError: If protocol version is not supported
    """
    # Check required fields
    if "version" not in envelope:
        raise InvalidMessageError("Missing 'version' field in envelope")

    if envelope["version"] != PROTOCOL_VERSION:
        raise UnsupportedVersionError(
            f"Unsupported protocol version: {envelope['version']}", {"version": envelope["version"]}
        )

    if "type" not in envelope:
        raise InvalidMessageError("Missing 'type' field in envelope")

    valid_types = {
        "request",
        "response",
        "error",
        "heartbeat",
        "register",
        "unregister",
        "stream_chunk",
        "stream_end",
    }
    if envelope["type"] not in valid_types:
        raise InvalidMessageError(
            f"Invalid message type: {envelope['type']}", {"type": envelope["type"]}
        )

    if "id" not in envelope:
        raise InvalidMessageError("Missing 'id' field in envelope")

    if "payload" not in envelope:
        raise InvalidMessageError("Missing 'payload' field in envelope")


def encode_bytes(envelope: dict[str, Any]) -> bytes:
    """Encode an envelope to bytes for transmission.

    Args:
        envelope: Envelope dictionary to encode

    Returns:
        UTF-8 encoded JSON bytes
    """
    return json.dumps(envelope).encode("utf-8")


def decode_bytes(data: bytes) -> dict[str, Any]:
    """Decode bytes to an envelope dictionary.

    Args:
        data: UTF-8 encoded JSON bytes

    Returns:
        Envelope dictionary

    Raises:
        MalformedPayloadError: If data cannot be decoded
    """
    try:
        decoded_data: dict[str, Any] = json.loads(data.decode("utf-8"))
        validate_envelope(decoded_data)
        return decoded_data
    except json.JSONDecodeError as e:
        raise MalformedPayloadError(f"Failed to decode JSON: {e}") from e
    except UnicodeDecodeError as e:
        raise MalformedPayloadError(f"Failed to decode UTF-8: {e}") from e

File: agenkit-ts/src/transports/codec.ts
/**
 * Message serialization and deserialization for protocol adapter.
 *
 * Provides encoding/decoding functions for Message objects and protocol envelopes.
 */

import { randomUUID } from 'crypto';
import { Message, ToolResult } from '../core/interfaces';
import {
  InvalidMessageError,
  MalformedPayloadError,
  UnsupportedVersionError,
} from './errors';

export const PROTOCOL_VERSION = '1.0';

/**
 * Protocol envelope for request/response communication.
 */
export interface ProtocolEnvelope {
  version: string;
  type: 'request' | 'response' | 'error' | 'heartbeat' | 'register' | 'unregister' | 'stream_chunk' | 'stream_end';
  id: string;
  timestamp: string;
  payload: Record<string, unknown>;
}

/**
 * Encode a Message object to a dictionary for JSON serialization.
 *
 * @param message The Message to encode
 * @returns Dictionary representation of the message
 */
export function encodeMessage(message: Message): Record<string, unknown> {
  return {
    role: message.role,
    content: message.content,
    metadata: message.metadata || {},
    timestamp: message.timestamp || new Date().toISOString(),
  };
}

/**
 * Decode a dictionary to a Message object.
 *
 * @param data Dictionary representation of a message
 * @returns Decoded Message object
 * @throws MalformedPayloadError if message data is invalid
 */
export function decodeMessage(data: Record<string, unknown>): Message {
  try {
    // Parse timestamp if present, otherwise use current time
    let timestamp = data.timestamp as string | undefined;
    if (!timestamp) {
      timestamp = new Date().toISOString();
    }

    return {
      role: data.role as string,
      content: data.content,
      metadata: (data.metadata as Record<string, unknown>) || {},
      timestamp,
    };
  } catch (error) {
    throw new MalformedPayloadError(
      `Failed to decode message: ${error}`,
      { data },
    );
  }
}

/**
 * Encode a ToolResult object to a dictionary for JSON serialization.
 *
 * @param result The ToolResult to encode
 * @returns Dictionary representation of the tool result
 */
export function encodeToolResult(result: ToolResult): Record<string, unknown> {
  return {
    success: result.success,
    output: result.output,
    error: result.error,
    metadata: result.metadata || {},
  };
}

/**
 * Decode a dictionary to a ToolResult object.
 *
 * @param data Dictionary representation of a tool result
 * @returns Decoded ToolResult object
 * @throws MalformedPayloadError if tool result data is invalid
 */
export function decodeToolResult(data: Record<string, unknown>): ToolResult {
  try {
    return {
      success: data.success as boolean,
      output: data.output,
      error: data.error as string | undefined,
      metadata: (data.metadata as Record<string, unknown>) || {},
    };
  } catch (error) {
    throw new MalformedPayloadError(
      `Failed to decode tool result: ${error}`,
      { data },
    );
  }
}

/**
 * Create a protocol request envelope.
 *
 * @param method Method name ("process" or "stream")
 * @param agentName Name of the target agent (for agent requests)
 * @param payload Request payload
 * @returns Request envelope dictionary
 */
export function createRequestEnvelope(
  method: string,
  agentName?: string,
  payload?: Record<string, unknown>,
): ProtocolEnvelope {
  return {
    version: PROTOCOL_VERSION,
    type: 'request',
    id: randomUUID(),
    timestamp: new Date().toISOString(),
    payload: {
      method,
      ...(agentName ? { agent_name: agentName } : {}),
      ...(payload || {}),
    },
  };
}

/**
 * Create a protocol response envelope.
 *
 * @param requestId ID of the request being responded to
 * @param payload Response payload
 * @returns Response envelope dictionary
 */
export function createResponseEnvelope(
  requestId: string,
  payload: Record<string, unknown>,
): ProtocolEnvelope {
  return {
    version: PROTOCOL_VERSION,
    type: 'response',
    id: requestId,
    timestamp: new Date().toISOString(),
    payload,
  };
}

/**
 * Create a protocol error envelope.
 *
 * @param requestId ID of the request that failed
 * @param errorCode Error code
 * @param errorMessage Human-readable error message
 * @param errorDetails Additional error details
 * @returns Error envelope dictionary
 */
export function createErrorEnvelope(
  requestId: string,
  errorCode: string,
  errorMessage: string,
  errorDetails?: Record<string, unknown>,
): ProtocolEnvelope {
  return {
    version: PROTOCOL_VERSION,
    type: 'error',
    id: requestId,
    timestamp: new Date().toISOString(),
    payload: {
      error_code: errorCode,
      error_message: errorMessage,
      error_details: errorDetails || {},
    },
  };
}

/**
 * Create a protocol stream chunk envelope.
 *
 * @param requestId ID of the streaming request
 * @param message Message chunk payload
 * @returns Stream chunk envelope dictionary
 */
export function createStreamChunkEnvelope(
  requestId: string,
  message: Record<string, unknown>,
): ProtocolEnvelope {
  return {
    version: PROTOCOL_VERSION,
    type: 'stream_chunk',
    id: requestId,
    timestamp: new Date().toISOString(),
    payload: { message },
  };
}

/**
 * Create a protocol stream end envelope.
 *
 * @param requestId ID of the streaming request
 * @returns Stream end envelope dictionary
 */
export function createStreamEndEnvelope(requestId: string): ProtocolEnvelope {
  return {
    version: PROTOCOL_VERSION,
    type: 'stream_end',
    id: requestId,
    timestamp: new Date().toISOString(),
    payload: {},
  };
}

/**
 * Validate a protocol envelope.
 *
 * @param envelope Envelope to validate
 * @throws InvalidMessageError if envelope is invalid
 * @throws UnsupportedVersionError if protocol version is not supported
 */
export function validateEnvelope(envelope: Record<string, unknown>): void {
  // Check required fields
  if (!envelope.version) {
    throw new InvalidMessageError("Missing 'version' field in envelope");
  }

  if (envelope.version !== PROTOCOL_VERSION) {
    throw new UnsupportedVersionError(
      `Unsupported protocol version: ${envelope.version}`,
      { version: envelope.version },
    );
  }

  if (!envelope.type) {
    throw new InvalidMessageError("Missing 'type' field in envelope");
  }

  const validTypes = new Set([
    'request',
    'response',
    'error',
    'heartbeat',
    'register',
    'unregister',
    'stream_chunk',
    'stream_end',
  ]);

  if (!validTypes.has(envelope.type as string)) {
    throw new InvalidMessageError(
      `Invalid message type: ${envelope.type}`,
      { type: envelope.type },
    );
  }

  if (!envelope.id) {
    throw new InvalidMessageError("Missing 'id' field in envelope");
  }

  if (!envelope.payload) {
    throw new InvalidMessageError("Missing 'payload' field in envelope");
  }
}

/**
 * Encode an envelope to bytes for transmission.
 *
 * @param envelope Envelope dictionary to encode
 * @returns UTF-8 encoded JSON bytes
 */
export function encodeBytes(envelope: ProtocolEnvelope): Buffer {
  return Buffer.from(JSON.stringify(envelope), 'utf-8');
}

/**
 * Decode bytes to an envelope dictionary.
 *
 * @param data UTF-8 encoded JSON bytes
 * @returns Envelope dictionary
 * @throws MalformedPayloadError if data cannot be decoded
 */
export function decodeBytes(data: Buffer): ProtocolEnvelope {
  try {
    const envelope = JSON.parse(data.toString('utf-8')) as Record<string, unknown>;
    validateEnvelope(envelope);
    return envelope as ProtocolEnvelope;
  } catch (error) {
    if (error instanceof MalformedPayloadError ||
        error instanceof InvalidMessageError ||
        error instanceof UnsupportedVersionError) {
      throw error;
    }
    throw new MalformedPayloadError(`Failed to decode envelope: ${error}`);
  }
}
